ARROW-11651: [Rust][DataFusion] Implement Postgres String Functions: Length Functions

seddonm1 opened the pull request:
Splitting up https://github.com/apache/arrow/pull/9243

This implements the following functions:

- String functions
  - [x] bit_Length
  - [x] char_length
  - [x] character_length
  - [x] length
  - [x] octet_length

jorgecarleitao commented on rust/arrow/src/compute/kernels/bit_length.rs:
```diff
@@ -0,0 +1,208 @@
+};
+use std::sync::Arc;
+
+fn bit_length_string<OffsetSize>(array: &Array, data_type: DataType) -> ArrayRef
+where
+    OffsetSize: OffsetSizeTrait,
+{
+    // note: offsets are stored as u8, but they can be interpreted as OffsetSize
+    let offsets = &array.data_ref().buffers()[0];
+    // this is a 30% improvement over iterating over u8s and building OffsetSize, which
+    // justifies the usage of `unsafe`.
+    let slice: &[OffsetSize] =
+        &unsafe { offsets.typed_data::<OffsetSize>() }[array.offset()..];
+
+    let bit_size = OffsetSize::from_usize(8).unwrap();
+    let lengths = slice
+        .windows(2)
+        .map(|offset| (offset[1] - offset[0]) * bit_size);
+
+    // JUSTIFICATION
+    //  Benefit
+    //      ~60% speedup
+    //  Soundness
+    //      `values` is an iterator with a known size.
+    let buffer = unsafe { Buffer::from_trusted_len_iter(lengths) };
+
```
Isn't this a copy of length, with a single difference, that we multiply the lengths by `bit_size`?

If that is the case, can we create a generic or something that avoids code duplication? Copy-pasting code like this is IMO an unnecessary maintenance burden.

seddonm1 replied:
yes i can have a look at doing that.

seddonm1 replied:
I have created a macro allowing just the core function to be replaced.

jorgecarleitao commented on rust/datafusion/src/physical_plan/functions.rs:
```diff
@@ -617,48 +670,137 @@ mod tests {
+    /// $ARRAY_TYPE is the column type after function applied
+    macro_rules! test_function {
+        ($FUNC:ident, $ARGS:expr, $EXPECTED:expr, $EXPECTED_TYPE:ty, $DATA_TYPE: ident, $ARRAY_TYPE:ident) => {
+            println!("{:?}", BuiltinScalarFunction::$FUNC);
```
is this `println!` intended?

seddonm1 replied:
it is just there to help debugging but happy to remove

seddonm1 replied:
Removed.

jorgecarleitao commented on rust/datafusion/src/physical_plan/string_expressions.rs:
```diff
@@ -115,6 +119,40 @@ where
+
+/// Returns number of characters in the string.
+/// character_length_i64('josé') = 4
+pub fn character_length_i64(args: &[ArrayRef]) -> Result<ArrayRef> {
```
I think that this code can be reduced to a single generic. Something like this (untested):

```rust
pub fn character_length<T: ArrowNativeType>(args: &[ArrayRef]) -> Result<ArrayRef> 
where 
T::Native: StringOffsetSizeTrait {
    let string_array: &GenericStringArray<T::Native> = args[0]
        .as_any()
        .downcast_ref::<GenericStringArray<T::Native>>()
        .unwrap();
    
     string_array
        .iter()
        .map(|x| x.map(|x: &str| T::Native::from_usize(x.graphemes(true).count())).unwrap())
        .collect()
}
```

seddonm1 replied:
Thanks I will have a go. There are a couple more instances this generic from_usize could help.

seddonm1 replied:
Wow. Impressive you can come up with this in your head. I have incorporated this code.

jorgecarleitao reviewed (commented):
Thanks a lot @seddonm1 . This looks great to me.

I left two comments that imo should be addressed related to reducing the maintenance cost by DRY.

seddonm1 commented:
Thanks @jorgecarleitao . I will update soon.

seddonm1 commented:
@jorgecarleitao Thanks for the review. I have addressed all of your points.

The `T::Native: StringOffsetSizeTrait` will be useful in some of the other string functions so thanks 🙏

jorgecarleitao commented:
Thanks a lot, @seddonm1 . I went through this PR again and I pull the code locally to understand the reason for the macro.

I now get it that it is not easy to write this. :) I PRed something to your branch with a suggestion to reduce the risk of using that function (which is unsafe due to the `typed`) on non-string arrays.

seddonm1 commented:
@jorgecarleitao Sorry I updated this but forgot to update you. Once this is merged I can do more of the big Postgres string functions PRs.

alamb commented:
The integration test failure appears unrelated https://github.com/apache/arrow/pull/9509/checks?check_run_id=1921983282

I will restart it to try and get a clean run

```

[INFO] Final Memory: 176M/780M
[INFO] ------------------------------------------------------------------------
Error:  Failed to execute goal org.xolstice.maven.plugins:protobuf-maven-plugin:0.5.0:compile (src) on project flight-core: Execution src of goal org.xolstice.maven.plugins:protobuf-maven-plugin:0.5.0:compile failed: Plugin org.xolstice.maven.plugins:protobuf-maven-plugin:0.5.0 or one of its dependencies could not be resolved: Could not transfer artifact org.apache.maven.plugin-tools:maven-plugin-annotations:jar:3.4 from/to central (https://repo.maven.apache.org/maven2): Connection reset -> [Help 1]
Error:  
Error:  To see the full stack trace of the errors, re-run Maven with the -e switch.
Error:  Re-run Maven using the -X switch to enable full debug logging.
Error:  
Error:  For more information about the errors and possible solutions, please read the following articles:
Error:  [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginResolutionException
Error:  
Error:  After correcting the problems, you can resume the build with the command
Error:    mvn <goals> -rf :flight-core
Destroying 3 processes
Destroying process..
Destroying process..
Destroying process..
```

alamb reviewed (approved):
Nice work @seddonm1  -- I think this is ready to go once it is green

alamb commented:
@seddonm1  we are so close -- now there is a clippy error: https://github.com/apache/arrow/pull/9509/checks?check_run_id=1922005826

```
ror: this function's return value is unnecessarily wrapped by `Result`
   --> arrow/src/compute/kernels/length.rs:86:1
    |
86  | / fn bit_length_impl<O: StringOffsetSizeTrait, T: ArrowPrimitiveType>(
87  | |     array: &dyn Array,
88  | | ) -> Result<ArrayRef>
89  | | where
...   |
99  | |     }))
100 | | }
    | |_^
    |
    = note: `-D clippy::unnecessary-wraps` implied by `-D warnings`
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_wraps
help: remove `Result` from the return type...
    |
88  | ) -> std::sync::Arc<(dyn array::array::Array + 'static)>
    |      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: ...and change the returning expressions
    |
97  |     unary_offsets_string::<O, _>(array, T::DATA_TYPE, |x| {
98  |         x * bits_in_bytes
99  |     })
    |

error: aborting due to previous error

error: could not compile `arrow`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: this function's return value is unnecessarily wrapped by `Result`
   --> arrow/src/compute/kernels/length.rs:86:1
    |
86  | / fn bit_length_impl<O: StringOffsetSizeTrait, T: ArrowPrimitiveType>(
87  | |     array: &dyn Array,
88  | | ) -> Result<ArrayRef>
89  | | where
...   |
99  | |     }))
100 | | }
    | |_^
    |
    = note: `-D clippy::unnecessary-wraps` implied by `-D warnings`
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_wraps
help: remove `Result` from the return type...
    |
88  | ) -> std::sync::Arc<(dyn array::array::Array + 'static)>
    |      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
help: ...and change the returning expressions
    |
97  |     unary_offsets_string::<O, _>(array, T::DATA_TYPE, |x| {
98  |         x * bits_in_bytes
99  |     })
    |
```

seddonm1 commented:
@alamb done and ready. Thanks to @jorgecarleitao for his elite knowledge of trait implementations.

jorgecarleitao reviewed (approved):
Thanks a lot, @seddonm1 ! Ready to go from my side.

alamb commented:
Thanks again @seddonm1  -- this is epic work
